OAuth 2.0 token exchange docs (#5181)
ydb-platform/ydb: ydb/docs/en/core/reference/ydb-sdk/_includes/auth.md

2. Otherwise, if the value of the `YDB_ANONYMOUS_CREDENTIALS` environment variable is set to 1, the anonymous authentication mode is used.
3. Otherwise, if the value of the `YDB_METADATA_CREDENTIALS` environment variable is set to 1, the **Metadata** authentication mode is used.
4. Otherwise, if the value of the `YDB_ACCESS_TOKEN_CREDENTIALS` environment variable is set, the **Access token** authentication mode is used, where the this variable value is passed.

> 5. Otherwise, if the value of the `YDB_OAUTH2_KEY_FILE` environment variable is set, the **OAuth 2.0 token exchange** authentication mode is used and the parameters are taken from the [json-file](#oauth2-key-file-format) whose name is specified in this variable.

Reviewer: ```suggestion
5. Otherwise, if the value of the `YDB_OAUTH2_KEY_FILE` environment variable is set, the **OAuth 2.0 token exchange** authentication mode is used, and the parameters are taken from the [JSON file](#oauth2-key-file-format) specified in this variable.
```

6. Otherwise, the **Metadata** authentication mode is used.

If the last step of the algorithm is selecting the **Metadata** mode, you can deploy a working application on VMs and in {{ yandex-cloud }} Cloud Functions without setting any environment variables.

> ## File format that is used for OAuth 2.0 token exchange authentication mode parameters {#oauth2-key-file-format}

Reviewer: ```suggestion
## File format for OAuth 2.0 token exchange authentication mode parameters {#oauth2-key-file-format}
```

> ## File format that is used for OAuth 2.0 token exchange authentication mode parameters {#oauth2-key-file-format}

Reviewer: This section needs some examples. Otherwise, there's no chance to write such a file from scratch.

> Description of fields of json-file with **OAuth 2.0 token exchange** authentication mode parameters.
> `creds_json` is json which holds parameters for token that is exchanged into access token. Two type are supported: `JWT` and `FIXED`,

Reviewer: ```suggestion
Description of fields of JSON file with **OAuth 2.0 token exchange** authentication mode parameters. The set of fields depends on the original token type, `JWT` and `FIXED`.

In the table below, `creds_json` means a JSON with parameters for exchanging the original token for an access token. 
```

| Field | Type | Description | Default value/optionality |
|:-----:|:----:|:-----------:|:-------------------------:|

> |grant-type|string|Grant type|`urn:ietf:params:oauth:grant-type:token-exchange`|
> |res|string|Resource|optional|
> |aud|string \| list of strings|Audience option for [token exchange request](https://www.rfc-editor.org/rfc/rfc8693)|optional|
> |scope|string \| list of strings|Scope|optional|
> |requested-token-type|string|Requested token type|`urn:ietf:params:oauth:token-type:access_token`|
> |subject-credentials|creds_json|Subject credentials|optional|
> |actor-credentials|creds_json|Actor credentials|optional|
> |token-endpoint|string|Token endpoint. In case of {{ ydb-short-name }} CLI it is overriden with `--iam-endpoint` option|optional|
> |**Description of fields of creds_json (JWT)**|||||
> |type|string|Token source type. Set `JWT`||
> |alg|string|Algorithm for JWT signature. Supported algorithms: ES256, ES384, ES512, HS256, HS384, HS512, PS256, PS384, PS512, RS256, RS384, RS512||
> |private-key|string|(Private) key in PEM format for JWT signature||
> |kid|string|`kid` JWT standard claim (key id)|optional|
> |iss|string|`iss` JWT standard claim (issuer)|optional|
> |sub|string|`sub` JWT standard claim (subject)|optional|
> |aud|string|`aud` JWT standard claim (audience)|optional|
> |jti|string|`jti` JWT standard claim (JWT id)|optional|
> |ttl|string|JWT token TTL|`1h`|

Reviewer: ```suggestion
|`grant-type`|string|Grant type|`urn:ietf:params:oauth:grant-type:token-exchange`|
|`res`|string|Resource|optional|
|`aud`|string \| list of strings|Audience option for [token exchange request](https://www.rfc-editor.org/rfc/rfc8693)|optional|
|`scope`|string \| list of strings|Scope|optional|
|`requested-token-type`|string|Requested token type|`urn:ietf:params:oauth:token-type:access_token`|
|`subject-credentials`|creds_json|Subject credentials|optional|
|`actor-credentials`|creds_json|Actor credentials|optional|
|`token-endpoint`|string|Token endpoint. In the case of {{ ydb-short-name }} CLI, it is overridden by the `--iam-endpoint` option.|optional|
|**Description of fields of creds_json (JWT)**|||||
|`type`|string|Token source type. Set `JWT`||
|`alg`|string|Algorithm for JWT signature. Supported algorithms: ES256, ES384, ES512, HS256, HS384, HS512, PS256, PS384, PS512, RS256, RS384, RS512||
|`private-key`|string|(Private) key in PEM format for JWT signature||
|`kid`|string|`kid` JWT standard claim (key id)|optional|
|`iss`|string|`iss` JWT standard claim (issuer)|optional|
|`sub`|string|`sub` JWT standard claim (subject)|optional|
|`aud`|string|`aud` JWT standard claim (audience)|optional|
|`jti`|string|`jti` JWT standard claim (JWT id)|optional|
|`ttl`|string|JWT token TTL|`1h`|
```

> |**Description of fields of creds_json (FIXED)**|||||
> |type|string|Token source type. Set `FIXED`||
> |token|string|Token value||
> |token-type|string|Token type value. It will become subject_token_type/actor_token_type parameter in [token exchange request](https://www.rfc-editor.org/rfc/rfc8693)||

Reviewer: ```suggestion
|**Description of fields of `creds_json` (FIXED)**|||||
|`type`|string|Token source type. Set `FIXED`||
|`token`|string|Token value||
|`token-type`|string|Token type value. It will become `subject_token_type/actor_token_type` parameter in [token exchange request](https://www.rfc-editor.org/rfc/rfc8693).||
```
